File: swift/DesignPattern.playground/Pages/FactoryMethodPattern.xcplaygroundpage/Sources/EUBeverageFactory.swift
import Foundation

open class EUBeverageFactory: BeverageFactory {
    
    public init() {}
    
    public func createBeverage(beverageName: String) -> Beverage? {
        var beverage: Beverage?
        
        switch beverageName {
        case "black tea":
            beverage = EarlGreyBlackTea()
        case "green tea":
            beverage = SenchaGreenTea()
        default:
            break
        }
        
        return beverage
    }
}

File: kotlin/DesignPattern/src/main/kotlin/factorymethodpattern/source/EUBeverageFactory.kt
package factorymethodpattern.source

class EUBeverageFactory: BeverageFactory {
    override fun createBeverage(beverageName: String): Beverage? {
        return when (beverageName) {
            "black tea" -> EarlGreyBlackTea()
            "green tea" -> SenchaGreenTea()
            else -> null
        }
    }
}